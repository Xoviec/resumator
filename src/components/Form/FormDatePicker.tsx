import { VoidFunctionComponent } from "react";
import { Controller, useFormContext } from "react-hook-form";
import { DesktopDatePicker, DatePickerProps } from "@mui/lab";
import { TextField } from "@mui/material";
import AdapterDateFns from "@mui/lab/AdapterDateFns";
import LocalizationProvider from "@mui/lab/LocalizationProvider";

type FormDatePickerProps = Omit<
  DatePickerProps,
  "value" | "onChange" | "renderInput"
> & {
  name: string;
};

export const FormDatePicker: VoidFunctionComponent<FormDatePickerProps> = ({
  name,
  label,
  inputFormat = "MM-yyyy",
  views = ["year", "month"],
  ...props
}) => {
  const { control } = useFormContext();

  return (
    <LocalizationProvider dateAdapter={AdapterDateFns}>
      <Controller
        defaultValue={null}
        control={control}
        name={name!}
        render={({ value, onChange }) => (
          <DesktopDatePicker
            inputFormat={inputFormat}
            value={value}
            onChange={onChange}
            renderInput={(textFieldProps) => (
              <TextField
                name={name}
                label={label}
                size="small"
                sx={{ flex: 1 }}
                {...textFieldProps}
              />
            )}
            views={views}
            {...props}
          />
        )}
      />
    </LocalizationProvider>
  );
};
