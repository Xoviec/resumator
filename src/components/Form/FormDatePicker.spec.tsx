import { fireEvent, render, screen } from "@testing-library/react";
import MockDate from "mockdate";
import { FunctionComponent } from "react";
import { FieldValues, FormProvider, useForm, UseFormReturn } from "react-hook-form";
import { FormDatePicker } from "./FormDatePicker";

beforeEach(() => {
  jest.resetAllMocks();
});

afterEach(() => {
  MockDate.reset();
});

describe("FormDatePicker", () => {
  it("should render correctly", () => {
    MockDate.set("2021-11-22T00:00:00.000Z");

    const FormProviderWrapper: FunctionComponent = ({ children }) => {
      const form = useForm();

      return <FormProvider {...form}>{children}</FormProvider>;
    };

    render(
      <FormProviderWrapper>
        <FormDatePicker name="test-form-name" label="test-form-label" />
      </FormProviderWrapper>
    );

    expect(document.body).toMatchSnapshot();
  });

  it("should update the value when selected", () => {
    MockDate.set("2021-11-03T00:00:00.000Z");

    let form: UseFormReturn<FieldValues, Record<string, unknown>> = {} as any;
    const fieldName = "test-form-name";

    const FormProviderWrapper: FunctionComponent<any> = ({ children, onChange }) => {
      form = useForm();

      return <FormProvider {...form}>{children}</FormProvider>;
    };

    render(
      <FormProviderWrapper>
        <FormDatePicker name={fieldName} label="test-form-label" />
      </FormProviderWrapper>
    );

    fireEvent.change(screen.getByRole("textbox"), {
      target: { value: "05-2021" },
    });

    expect(form.getValues()).toEqual({
      "test-form-name": new Date("2021-05-01T00:00:00.000Z"),
    });
  });
});
