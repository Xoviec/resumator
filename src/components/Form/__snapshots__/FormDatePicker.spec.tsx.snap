// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormDatePicker should render correctly 1`] = `
<body>
  <div>
    <div
      class="MuiFormControl-root MuiTextField-root css-1vs1jgp-MuiFormControl-root-MuiTextField-root"
      role="input"
    >
      <label
        class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for="mui-1"
        id="mui-1-label"
      >
        test-form-label
      </label>
      <div
        class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd css-19qh8xo-MuiInputBase-input-MuiOutlinedInput-input"
          id="mui-1"
          name="test-form-name"
          placeholder="mm-yyyy"
          type="text"
          value=""
        />
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeSmall css-1laqsz7-MuiInputAdornment-root"
        >
          <button
            aria-label="Choose date"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CalendarIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-1ftyaf0"
          >
            <span>
              test-form-label
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</body>
`;
